torchtitan/parallelisms/parallel_dims.py

```diff
@@ -18,36 +18,53 @@ class ParallelDims:
     pp: int
     world_size: int
     enable_loss_parallel: bool
-    dp_type: str
+    dp_type: str = "fsdp"
+    dp_replicate: int = 1  # Only used when dp_type is hsdp
 
     def __post_init__(self):
         self.dp_type = self.dp_type.lower()
         self._validate()
 
     def _validate(self):
-        dp, tp, pp = self.dp, self.tp, self.pp
+        dp, dp_replicate, tp, pp = self.dp, self.dp_replicate, self.tp, self.pp
         if dp == -1:
             self.dp = dp = self.world_size // (tp * pp)
         assert dp >= 1, dp
+        assert dp_replicate >= 1 and dp % dp_replicate == 0, (dp, dp_replicate)
         assert tp >= 1, tp
         assert pp >= 1, pp
         assert (
-            dp * tp * pp == self.world_size
+            dp * tp * pp == self.world_size,
         ), f"Invalid parallel dims: dp({dp}) * tp({tp}) * pp({pp}) != WORLD_SIZE({self.world_size})"
-        assert self.dp_type in ("fsdp", "ddp")
+        assert self.dp_type in ("fsdp", "ddp", "hsdp")
+        assert self.dp_type != "hsdp" or dp_replicate > 1, (self.dp_type, dp_replicate)
 
     def build_mesh(self, device_type):
         dims = []
         names = []
         for d, name in zip(
             [self.pp, self.dp, self.tp], ["pp", "dp", "tp"], strict=True
         ):
-            if d > 1:
+            if d <= 1:
+                continue
+
+            if name != "dp" or self.dp_replicate <= 1:
                 dims.append(d)
                 names.append(name)
+                continue
+
+            dp_shard = self.dp // self.dp_replicate
+            dims.extend([self.dp_replicate, dp_shard])
+            names.extend(["dp_replicate", "dp_shard"])
+
         logger.info(f"Building {len(dims)}-D device mesh with {names}, {dims}")
         names = tuple(names)
-        return init_device_mesh(device_type, dims, mesh_dim_names=names)
+        mesh = init_device_mesh(device_type, dims, mesh_dim_names=names)
+        # Create all the submesh here to ensure all required process groups are
+        # initialized
+        if self.dp_replicate > 1:
+            mesh["dp_replicate", "dp_shard"]._flatten()
```

Comment:
hmm, this initializes a new PG for "dp" but does not name it? so later users will be able to access this initialized PG by doing the same slice,flatten combo and hitting a cache?  The UX feels a little "surprising" to me

Reply:
Please see the discussion in https://github.com/pytorch/pytorch/pull/132779. Currently, the name is automatically created.

Reply:
cc., @wz337, @wanchaol

Reply:
So this `_flatten()` initializes the PGs and the later `_flatten()` in `train.py` just returns it?

curious if there is any explicit motivation for eager initializing the PG

Reply:
@wconstab suggested that we should initialize all PGs in the same place. Since initializing PG requires all ranks participation, this can reduce the possible deadlock and easier for users to debug, though this is not required.

Reply:
> hmm, this initializes a new PG for "dp" but does not name it? so later users will be able to access this initialized PG by doing the same slice,flatten combo and hitting a cache? The UX feels a little "surprising" to me

The initial design is that the user must call flatten() explicitly so we will do the pg initialization for the new flatten mesh dim. Later, the user could access the flattened mesh via slicing.

```
mesh_3d = init_device_mesh("cuda", (2,2,2), ("replicate", "shard", "cp"))
dp_shard_cp_mesh = mesh_3d["shard", "cp"].flatten()
# then you can do the following slicing, since "shard_cp" flatten mesh has been created.
fully_shard(model, device_mesh=mesh3d["replicate", "shard_cp"]
```

As no naming is currently allowed, the new flatten mesh_dim_name is automatically set to concatenate the two original mesh_dim_name by "_". However, now I do see a problem if that the original mesh_dim_name has already contained "_", for example", "dp_replicate" or "dp_shard", this won't work. 

cc. @wanchaol